bin_opt/optimize_binning_bbWW.py

```diff
@@ -0,0 +1,215 @@
+                        custom_totals.append(curr_integral_signal)
+                        print("At bin ", left_edge, " integral is ", custom_totals[-1])
+                        right_edge = left_edge-1
+                        big_bin_counter+=1
+                        break
+
+            print("Found the bin edge set, it is ", custom_binning)
+            print("With totals per bin as ", custom_totals)
+
+            if custom_totals[-1] < 0.2*bin_content_goal:
+                print("Final bin was very very small, auto merged!")
+                del custom_binning[-2]
+                print("New binning is ", custom_binning)
+
+            custom_binning.sort()
+            bins_dict[dict_key] = custom_binning
+
+            print("Final dict binnings are ", bins_dict)
+
+
+
+            json_dict = convert_to_json(bins_dict, params)
+
+            with open(f'tmp_binning_{sig_string}.json', 'w') as f:
+                json.dump(json_dict, f)
+
+
+            #Create the datacards
+            ps_call(f"python3 ../dc_make/create_datacards.py --input /eos/user/d/daebi/bbWW_shape_files --output tmp_datacards_{sig_string} --config ../config/x_hh_bbww_run3.yaml --hist-bins ./tmp_binning_{sig_string}.json --param_values {mass_list}", shell=True)
+            #Find out where the limits will be saved
+            output = ps_call(f"law run MergeResonantLimits --version dev --datacards 'tmp_datacards_{sig_string}/*.txt' --print-output 0,False", shell=True, catch_stdout=True, split="\n")[1]
+            limit_file_name = output[-2]
+            #Run the limit calculation
+            ps_call(f"law run MergeResonantLimits --version dev --datacards 'tmp_datacards_{sig_string}/*.txt' --remove-output 3,a,y", shell=True)
+
+            #Now load the output numpy array and get value [1] (mass, val, up1, down1, up2, down2)
+            print(np.load(limit_file_name).files)
+            data = np.load(limit_file_name)
+            exp_limit = data[data.files[0]][0][1]
+
+            limit_history[dict_key]['nBins'].append(nbins)
+            limit_history[dict_key]['limits'].append(exp_limit)
+            limit_list.append(exp_limit)
+
+            #If adding a bin does not improve (limit stays the same) then just move on
+            if (exp_limit < best_limit) or (nbins == 1):
+                best_limit = exp_limit
+                best_bins = bins_dict.copy()
+
+            else:
+                print(f"Did not improve limits at nbins {nbins}, best limit was {best_limit}, and current was {exp_limit}")
+                #Need to replace the bin dict because it now has the 'worse' new binning
+                bins_dict = best_bins.copy()
+                break
+
+
+
+
+
+    print("Finished the limit scan, lets look at the history")
+    print(limit_history)
+
+    print(limit_list)
+    plt.plot(limit_list)
+    #plt.show()
+    plt.yscale('log')
+    plt.savefig(f"binopt_history_{sig_string}.pdf")
+    plt.close()
+
+
+
+
+if __name__ == '__main__':
```

Comment:
add inputs, outputs, list of processes, max number of bins etc. as command line arguments

Reply:
Added many arguments in newest version including a call example for command line args